fiftyone/core/plots/utils.py

```diff
@@ -5,46 +5,336 @@
+
+    raw_y_values = not (etau.is_str(y) or _is_expr(y))
+
+    if raw_y_values:
+        is_frames = y and etau.is_container(y[0])
+    elif samples is not None and samples._contains_videos():
+        is_frames = foe.is_frames_expr(y)
+    else:
+        is_frames = False
+
+    if is_frames and link_field is None:
+        link_field = "frames"
+
+    if ids is None and samples is not None:
+        ref = y if raw_y_values else None
+        ids = _get_ids(
+            samples, link_field=link_field, ref=ref, is_frames=is_frames
+        )
+
+    x = _parse_values(
+        x,
+        "x",
+        samples=samples,
+        ids=ids,
+        link_field=link_field,
+        is_frames=is_frames,
+    )
+
+    y = _parse_values(
+        y,
+        "y",
+        samples=samples,
+        ids=ids,
+        link_field=link_field,
+        is_frames=is_frames,
+    )
+
+    sizes = _parse_values(
+        sizes,
+        "sizes",
+        samples=samples,
+        ids=ids,
+        link_field=link_field,
+        ref=y,
+        is_frames=is_frames,
+    )
+
+    if x is None:
+        if is_frames:
+            x = [np.arange(1, len(yi) + 1) for yi in y]
+        else:
+            x = np.arange(1, len(y) + 1)
+
+    if is_frames and x and not etau.is_container(x[0]):
+        x = [x] * len(y)
+
+    if is_frames:
+        if sizes is None:
+            sizes = [None] * len(y)
+
+        if ids is None:
+            ids = [None] * len(y)
+
+        if labels is None:
+            labels = [str(i) for i in range(1, len(y) + 1)]
+        else:
+            labels = _parse_values(labels, "labels", samples=samples)
+    else:
+        x = [x]
+        y = [y]
+        ids = [ids]
+        sizes = [sizes]
+        labels = [labels]
+
+    return x, y, ids, link_field, labels, sizes, is_frames
+
+
+def best_fit_line(points, label=None):
+    x = points[:, 0]
+    y = points[:, 1]
+
+    model = skl.LinearRegression()
+    model.fit(x[:, np.newaxis], y)
+
+    xline = np.array([x.min(), x.max()])
+    yline = model.predict(xline[:, np.newaxis])
 
-    Args:
-        icon_bytes:
+    if label is None:
+        r2_score = skm.r2_score(x, y, sample_weight=None)
+        label = "r^2: %0.3f" % r2_score
 
-    Returns:
-        the numpy image
-    """
+    return xline, yline, label
+
+
+def _is_expr(arg):
+    return isinstance(arg, (foe.ViewExpression, dict))
+
+
+def _parse_values(
+    values,
+    parameter,
+    samples=None,
+    ids=None,
+    link_field=None,
+    ref=None,
+    is_frames=False,
+):
+    if values is None:
+        return None
+
+    if etau.is_str(values) or isinstance(values, foe.ViewExpression):
+        if samples is None:
+            raise ValueError(
+                "You must provide `samples` in order to extract field values "
+                "for the `%s` parameter" % parameter
+            )
+
+        if ids is not None and not is_frames:
+            values = samples._get_values_by_id(
```

Comment:
All of the refactoring in the plotting methods was to achieve this line: when the user has provided IDs and is asking to pull values from a dataset via path/expression, use `_get_values_by_id()` to ensure that the correct values are pulled, even if `samples` doesn't correspond 1-1 with other data that may have been provided.